### PR TITLE
Fix tasks being wrongly skipped by schedule_after_task_execution

### LINKED ISSUE #19222: none_failed_min_one_success trigger rule not working with BranchPythonOperator in certain cases.
### Apache Airflow version

2.2.0 (latest released)

### Operating System

CentOS Linux 7 (Core)

### Versions of Apache Airflow Providers

apache-airflow-providers-ftp 2.0.1
apache-airflow-providers-http 2.0.1
apache-airflow-providers-imap 2.0.1
apache-airflow-providers-sqlite 2.0.1

### Deployment

Other

### Deployment details

centos 7.3
postgres - 12.2

### DAG 1

In this DAG, I am expecting task_6 to run even when one of task_4 or task_5 get skipped. But, as soon as task_5 skips, task_6 also gets skipped. Ideally, task_6 should have waited for task_4 to finish, then only it should have taken the decision whether to run or skip.
```
import datetime as dt
import time

from airflow import DAG
from airflow.operators.python_operator import BranchPythonOperator
from airflow.operators.dummy_operator import DummyOperator


dag = DAG(
    dag_id="test_non_failed_min_one_success",
    schedule_interval="@once",
    start_date=dt.datetime(2019, 2, 28),
)


def sleep(seconds, return_val=None):
    time.sleep(seconds)
    return return_val


op1 = DummyOperator(task_id="task_1", dag=dag)
op2 = BranchPythonOperator(
    task_id="task_2", python_callable=sleep, op_args=[30, ["task_4"]], dag=dag
)
op3 = BranchPythonOperator(task_id="task_3", python_callable=sleep, op_args=[10, []], dag=dag)
op4 = DummyOperator(task_id="task_4", dag=dag)
op5 = DummyOperator(task_id="task_5", dag=dag)
op6 = DummyOperator(task_id="task_6", dag=dag, trigger_rule="none_failed_min_one_success")

op1 >> [op2, op3]
op2 >> op4
op3 >> op5
[op4, op5] >> op6
```

image
image

### DAG 2

This is just a modification of DAG 1 where I have created two more dummy tasks between task_5 and task_6. Now, I get the desired behaviour, i.e. task_6 waits for both task_4 and dummy_2 to finish before taking the decision of whether to run or skip.

```
import datetime as dt
import time

from airflow import DAG
from airflow.operators.python_operator import BranchPythonOperator
from airflow.operators.dummy_operator import DummyOperator


dag = DAG(
    dag_id="test_non_failed_min_one_success",
    schedule_interval="@once",
    start_date=dt.datetime(2019, 2, 28),
)


def sleep(seconds, return_val=None):
    time.sleep(seconds)
    return return_val


op1 = DummyOperator(task_id="task_1", dag=dag)
op2 = BranchPythonOperator(
    task_id="task_2", python_callable=sleep, op_args=[30, ["task_4"]], dag=dag
)
op3 = BranchPythonOperator(task_id="task_3", python_callable=sleep, op_args=[10, []], dag=dag)
op4 = DummyOperator(task_id="task_4", dag=dag)
op5 = DummyOperator(task_id="task_5", dag=dag)
dummy1 = DummyOperator(task_id="dummy1", dag=dag)
dummy2 = DummyOperator(task_id="dummy2", dag=dag)
op6 = DummyOperator(task_id="task_6", dag=dag, trigger_rule="none_failed_min_one_success")

op1 >> [op2, op3]
op2 >> op4
op3 >> op5
op5 >> dummy1 >> dummy2
[op4, dummy2] >> op6
```

image

image

### What you expected to happen

I expected task_6 in DAG 1 to wait for both the parent tasks to finish their runs and then run. This is because I have set trigger rule = "none_failed_min_one_success" in task_6. But, this trigger rule is not being respected in DAG 1.

### How to reproduce

The above code can be used to reproduce the issue.

### Anything else

This works fine in version 2.0.2 with trigger rule = "none_failed_or_skipped".

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by tanelk**
closes: #19222
Alternative to #22374

#22374 explains the issue well, but the aproach would limit the mini scheduler to the most basic trigger rules.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -1964,7 +1964,13 @@ def partial_subset(
             tasks, in addition to matched tasks.
         :param include_upstream: Include all upstream tasks of matched tasks,
             in addition to matched tasks.
+        :param include_direct_upstream: Include all tasks directly upstream of matched
+            and downstream (if include_downstream = True) tasks. Can not be use when
+            include_upstream = True - would be a noop with overhead.
```
Comment (tanelk) [inline airflow/models/dag.py]: Before it was "Include all tasks directly upstream of matched", now it has the "and downstream" part.

File: tests/jobs/test_local_task_job.py [hunk 2]
```diff
@@ -799,6 +800,121 @@ def test_mini_scheduler_works_with_wait_for_upstream(self, caplog, dag_maker):
...
+        [
+            (AirflowFailException(), TriggerRule.ALL_DONE),
+            (AirflowFailException(), TriggerRule.ALL_FAILED),
+            (AirflowSkipException(), TriggerRule.ALL_DONE),
+            (AirflowSkipException(), TriggerRule.ALL_SKIPPED),
+            (AirflowSkipException(), TriggerRule.NONE_FAILED_MIN_ONE_SUCCESS),
+        ],
+    )
+    @conf_vars({('scheduler', 'schedule_after_task_execution'): 'True'})
+    def test_mini_scheduler_works_with_skipped_and_failed(self, exception, trigger_rule, caplog, dag_maker):
```
Comment (tanelk) [inline tests/jobs/test_local_task_job.py]: These pass even without the PR - initially tried to recreate the issue with these, but now left just to improve coverage.

File: tests/jobs/test_local_task_job.py [hunk 3]
```diff
@@ -799,6 +800,121 @@ def test_mini_scheduler_works_with_wait_for_upstream(self, caplog, dag_maker):
...
+        "trigger_rule",
+        [
+            TriggerRule.ONE_SUCCESS,
+            TriggerRule.ALL_SKIPPED,
+            TriggerRule.NONE_FAILED,
+            TriggerRule.NONE_FAILED_MIN_ONE_SUCCESS,
+        ],
+    )
+    @conf_vars({('scheduler', 'schedule_after_task_execution'): 'True'})
+    def test_mini_scheduler_works_with_branch_python_operator(self, trigger_rule, caplog, dag_maker):
```
Comment (tanelk) [inline tests/jobs/test_local_task_job.py]: This fails without PR

File: airflow/models/dag.py [hunk 4]
```diff
@@ -1984,15 +1987,18 @@ def partial_subset(
                 also_include.extend(t.get_flat_relatives(upstream=False))
             if include_upstream:
                 also_include.extend(t.get_flat_relatives(upstream=True))
-            elif include_direct_upstream:
+
+        direct_upstream: List[Operator] = []
```
Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
        direct_upstreams: List[Operator] = []
```

File: airflow/models/dag.py [hunk 5]
```diff
@@ -1984,15 +1987,18 @@ def partial_subset(
...
                 upstream = (u for u in t.upstream_list if isinstance(u, (BaseOperator, MappedOperator)))
-                also_include.extend(upstream)
+                direct_upstream.extend(upstream)
 
         # Compiling the unique list of tasks that made the cut
         # Make sure to not recursively deepcopy the dag while copying the task
         dag.task_dict = {
             t.task_id: copy.deepcopy(t, {id(t.dag): dag})  # type: ignore
-            for t in matched_tasks + also_include
+            for t in matched_tasks + also_include + direct_upstream
```
Comment (uranusjr) [inline airflow/models/dag.py]: We should use `itertools.chain` here.

File: tests/jobs/test_local_task_job.py [hunk 6]
```diff
@@ -799,6 +800,121 @@ def test_mini_scheduler_works_with_wait_for_upstream(self, caplog, dag_maker):
...
+    @conf_vars({('scheduler', 'schedule_after_task_execution'): 'True'})
+    def test_mini_scheduler_works_with_skipped_and_failed(self, exception, trigger_rule, caplog, dag_maker):
+        """
+        In these cases D is running, at no decision can be made about C.
+        """
+
+        def raise_():
+            raise exception
+
+        session = settings.Session()
```
Comment (uranusjr) [inline tests/jobs/test_local_task_job.py]: You can use the `session` fixture here instead, which automatically calls `rollback` on test teardown.

File: tests/jobs/test_local_task_job.py [hunk 7]
```diff
@@ -799,6 +800,121 @@ def test_mini_scheduler_works_with_wait_for_upstream(self, caplog, dag_maker):
...
+        assert ti_b.state in (State.SKIPPED, State.UPSTREAM_FAILED)
+        assert ti_c.state == State.NONE
+        assert "0 downstream tasks scheduled from follow-on schedule" in caplog.text
+
+        failed_deps = list(ti_c.get_failed_dep_statuses(session=session))
+        assert len(failed_deps) == 1
+        assert failed_deps[0].dep_name == "Trigger Rule"
+        assert not failed_deps[0].passed
+
+        session.rollback()
```
Comment (uranusjr) [inline tests/jobs/test_local_task_job.py]: … because due to how tests run, this won’t be called if any of the asserts fail.

**Review by ashb — APPROVED** [review]
This might need optimizing later for larger dags that don't use anything but the standard trigger rule, but good for now.

(That is: if we are using standard ALL_SUCCESS trigger rule we could possibly not pull in the extra upstream tasks to check.)

**Pull request merged**